src/librustdoc/html/render/cache.rs

```diff
@@ -357,35 +374,35 @@ crate fn get_all_types<'tcx>(
     generics: &Generics,
     decl: &FnDecl,
     tcx: TyCtxt<'tcx>,
-) -> (Vec<(Type, ItemType)>, Vec<(Type, ItemType)>) {
-    let mut all_types = FxHashSet::default();
+) -> (Vec<TypeWithKind>, Vec<TypeWithKind>) {
+    let mut all_types = Vec::new();
+    let mut args = Vec::new();
     for arg in decl.inputs.values.iter() {
         if arg.type_.is_self_type() {
             continue;
         }
-        let mut args = FxHashSet::default();
+        args.clear();
         get_real_types(generics, &arg.type_, tcx, 0, &mut args);
         if !args.is_empty() {
-            all_types.extend(args);
+            all_types.extend(args.drain(..));
```

Comment:
Why did you move `let mut args` outside of the loop and add `.clear()` and `.drain(..)`? Is there a behavioral difference between the old and new code?

Reply:
There are two things happening here:

 * Before you couldn't have the same type multiple times (because of the hashset), which is problematic when you're looking for `u32, u32` for example (function taking two `u32` for example).
 * As for the `clear` and `drain`, it's simply to be sure that the `args` vector only contains what we're looking for and improve memory usage.

Reply:
Could you remove the second change and instead have the `let mut args = vec![];` inside the loop? I know it might not be as performant, but having the state persist across iterations makes this change harder to review. I'd rather not mix performance and behavior changes too.

Reply:
Is it ok if I split it into two commits?

Reply:
I'd rather it be a separate PR; I see some potential edge cases with it, and the slight performance improvement doesn't seem worth it to me (at least for now) for the additional code complexity.

Reply:
I'll add a FIXME then.

Reply:
Done!